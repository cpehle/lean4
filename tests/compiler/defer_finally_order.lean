/-! Test the execution order of defer vs finally -/

def test1 : IO Unit := do
  IO.println "=== Test 1: Defer and finally order ==="
  try
    IO.println "In try block"
    defer IO.println "Defer in try"
    IO.println "After defer declaration"
  catch _ =>
    IO.println "Should not reach catch"
  finally
    IO.println "In finally block"

def test2 : IO Unit := do
  IO.println "\n=== Test 2: Multiple defers with finally ==="
  try
    defer IO.println "Defer 1"
    defer IO.println "Defer 2"
    IO.println "Normal execution"
  finally
    IO.println "Finally block"

def test3 : IO Unit := do
  IO.println "\n=== Test 3: Defer, exception, and finally ==="
  try
    defer IO.println "Defer before exception"
    IO.println "About to throw"
    throw (IO.userError "Test exception")
  catch e =>
    IO.println s!"Caught: {e}"
  finally
    IO.println "Finally after exception"

def main : IO Unit := do
  test1
  test2
  test3
  IO.println "\n=== Tests completed ==="