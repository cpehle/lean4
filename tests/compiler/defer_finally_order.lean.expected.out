=== Test 1: Defer and finally order ===
In try block
After defer declaration
In finally block
Defer in try

=== Test 2: Multiple defers with finally ===
Normal execution
Finally block
Defer 1
Defer 2

=== Test 3: Defer, exception, and finally ===
About to throw
Caught: Test exception
Finally after exception
Defer before exception

=== Tests completed ===
